From c708ba935538edeb4477f404f9fdec5975958059 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Sun, 18 Jun 2017 09:00:16 +0300
Subject: [PATCH 1065/1345] fix: net: mvpp2: Add MAC reset and force link down
 to port en/dis

commit  793a2baab00fad58ac2403f5b090e5b29e63fc5a from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

This patch add MAC reset and force link down to port enable/disable.
Force link down and port in MAC reset are required during port
initialization procedure.
Missing MAC reset and force link down configurations could cause
MVPP2 RX FIFO overrun and GoP stuck.

Change-Id: Ie77ce3aa3d1b680a207f4a7de66e427f73b74df4
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/40553
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2.c |   16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2.c b/drivers/net/ethernet/marvell/mvpp2.c
index 6587808..9293de0 100644
--- a/drivers/net/ethernet/marvell/mvpp2.c
+++ b/drivers/net/ethernet/marvell/mvpp2.c
@@ -4262,6 +4262,14 @@ static void mvpp2_port_enable(struct mvpp2_port *port)
 		val |= MVPP2_GMAC_PORT_EN_MASK;
 		val |= MVPP2_GMAC_MIB_CNTR_EN_MASK;
 		writel(val, port->base + MVPP2_GMAC_CTRL_0_REG);
+
+		val = readl(port->base + MVPP2_GMAC_AUTONEG_CONFIG);
+		val &= ~(MVPP2_GMAC_FORCE_LINK_DOWN);
+		writel(val, port->base + MVPP2_GMAC_AUTONEG_CONFIG);
+
+		val = readl(port->base + MVPP2_GMAC_CTRL_2_REG);
+		val &= ~(MVPP2_GMAC_PORT_RESET_MASK);
+		writel(val, port->base + MVPP2_GMAC_CTRL_2_REG);
 	}
 }
 
@@ -4281,6 +4289,14 @@ static void mvpp2_port_disable(struct mvpp2_port *port)
 		val = readl(port->base + MVPP2_GMAC_CTRL_0_REG);
 		val &= ~(MVPP2_GMAC_PORT_EN_MASK);
 		writel(val, port->base + MVPP2_GMAC_CTRL_0_REG);
+
+		val = readl(port->base + MVPP2_GMAC_AUTONEG_CONFIG);
+		val |= MVPP2_GMAC_FORCE_LINK_DOWN;
+		writel(val, port->base + MVPP2_GMAC_AUTONEG_CONFIG);
+
+		val = readl(port->base + MVPP2_GMAC_CTRL_2_REG);
+		val |= MVPP2_GMAC_PORT_RESET_MASK;
+		writel(val, port->base + MVPP2_GMAC_CTRL_2_REG);
 	}
 }
 
-- 
1.7.9.5

